huaweicloud/resource_huaweicloud_apig_instance.go

```diff
@@ -0,0 +1,520 @@
+	}
+	if v, ok := d.GetOk("maintain_time"); ok {
+		startTime := v.(string)
+		opt.MaintainBegin = startTime
+		endTime, err := buildMaintainEndTime(startTime)
+		if err != nil {
+			return opt, err
+		}
+		opt.MaintainEnd = endTime
+	}
+
+	return opt, nil
+}
+
+func watiForApigInstanceV2TargetState(d *schema.ResourceData, client *golangsdk.ServiceClient, ref Refresh) error {
+	stateConf := &resource.StateChangeConf{
+		Pending: ref.Pending,
+		Target:  ref.Target,
+		Refresh: ApigInstanceV2StateRefreshFunc(client, d.Id()),
+		Timeout: ref.Timeout,
+		Delay:   ref.Delay,
+	}
+	if ref.MinTimeout != 0 {
+		stateConf.MinTimeout = ref.MinTimeout
+	} else {
+		stateConf.PollInterval = ref.PollInterval
+	}
+	_, err := stateConf.WaitForState()
+	return err
+}
+
+func ApigInstanceV2StateRefreshFunc(client *golangsdk.ServiceClient, id string) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		opt := instances.ListOpts{
+			Id: id,
+		}
+		// Some status cannot be read by GET method, just like 'Deleting'.
+		// GET method will link to other table (vpc) for query. The response time is not as good as the LIST method.
+		allPages, err := instances.List(client, opt).AllPages()
+		if err != nil {
+			return allPages, "", fmt.Errorf("Error getting APIG v2 dedicated instance by ID (%s): %s", id, err)
+		}
+		instances, err := instances.ExtractInstances(allPages)
+		if len(instances) == 0 {
+			return instances, "DELETED", nil
+		}
+		return instances[0], instances[0].Status, nil
+	}
+}
+
+func waitForApigInstanceCreateCompleted(d *schema.ResourceData, client *golangsdk.ServiceClient) error {
+	ref := Refresh{
+		Pending:      []string{"Creating", "Initing", "Installing", "Registering"},
+		Target:       []string{"Running"},
+		Delay:        30 * time.Second,
+		Timeout:      d.Timeout(schema.TimeoutCreate),
+		PollInterval: 10 * time.Second,
+	}
+	return watiForApigInstanceV2TargetState(d, client, ref)
+}
+
+func resourceApigInstanceV2Create(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*config.Config)
+	opts, err := buildApigInstanceParameters(d, config)
+	if err != nil {
+		return fmt.Errorf("Error craeting APIG v2 dedicated instance options: %s", err)
+	}
+	log.Printf("[DEBUG] Create APIG v2 dedicated instance options: %#v", opts)
+
+	client, err := config.ApigV2Client(GetRegion(d, config))
+	if err != nil {
+		return fmt.Errorf("Error creating HuaweiCloud APIG v2 client: %s", err)
+	}
+	v, err := instances.Create(client, opts).Extract()
+	if err != nil {
+		return fmt.Errorf("Error creating HuaweiCloud APIG v2 dedicated instance: %s", err)
+	}
+	d.SetId(v.Id)
+	err = waitForApigInstanceCreateCompleted(d, client)
+	if err != nil {
+		return fmt.Errorf("Error waiting for APIG v2 dedicated instance (%s) to become running: %s", d.Id(), err)
+	}
+	return resourceApigInstanceV2Read(d, meta)
+}
+
+func setApigAvailableZoneIds(d *schema.ResourceData, resp instances.Instance) error {
```

Comment:
please add more comments to explain why we should do this.

Reply:
the function description has been added.